### DISCUSSION TITLE
Obtaining ccHash from the Nook Windows store app

### DISCUSSION THREAD

**Discussion by fesiwi** [noDRM/DeDRM_tools#9] [Show and tell]
Experimental Python script to extract the Barnes and Noble Nook decryption key (ccHash) from the Nook Windows store app. If it proves useful perhaps it could be included into the DeDRM plugin. The actual script has been zipped for posting here.

Instructions:

1. Install calibre version 5 or later on your PC. (This Python script relies on calibre version 5 or later being installed to supply
        the correct libraries for its execution.)

2. Install the Nook app from: https://www.microsoft.com/en-us/p/nook-books-magazines-newspapers-comics/9wzdncrfj33h
        (Requires a recent Windows version.)

3. Sign in to the Nook app with your email and password and then close the app.

4. Run this script by entering at a Windows command prompt:
        calibre-debug -e nook-decoder.py

This procedure has been tested with Nook app version 1.11.0.4 under Windows 11.

[nook-decoder.zip](https://github.com/noDRM/DeDRM_tools/files/7740134/nook-decoder.zip)

**Comment by ElleKayEm** [comment]
Beautiful!  I look forward to trying this when I get a chance.

**Comment by noDRM** [comment]
Wow, that looks way easier than all the other current workarounds to acquire the key. 

I will give this a try and if it works without issues I will definitely add that code to the plugin. Thanks!

**Comment by ElleKayEm** [comment]
This worked perfectly for me!  Thanks so much.

**Comment by olderTechnology** [comment]
Thank you fesiwi! I spent days trying to figure out the new B&N storage in the summer. Finally gave up. Your script gave me my key and I am able to unlock Nook books again!

**Comment by noDRM** [comment]
Just tried to test this but I can't install the Nook app. Tested two Windows 10 machines with an up-to-date Windows (21H2), neither of these display the "Install" / "Download" button for the Nook app. Am I doing something wrong, or did they take the app down? Maybe they noticed this script already and are building a new version?

**Reply by ElleKayEm** [reply]
I already had the Nook app installed on the machine I tested this with the other day.  Just tried on another one and was able to install it.  Clicked the link above and there was a "Get" button.  Once again Nook decoder retrieved the key.

**Reply by noDRM** [reply]
Huh. Interesting. Apparently the app is only available in certain countries. Once I switched my Windows install's region to USA I managed to install it.

**Comment by noDRM** [comment]
I've just included that code in the plugin (including support for Python2 / Calibre < 5). When you click the "+" button for a new B&N key, you'll be asked if you want to use the "old" way (username and password / CC number), if you want to extract data from the installed Windows app (this uses your code), or if you want to grab the key from dumped Android app data (this is your code adapted to the Android app, which works similarly). 

Support for the key retrieval method (entering mail address and password, then downloading the ccHash directly from B&N) has been removed as the necessary server seems to have been shut down long ago.

If you want to test without waiting for the next release, you can click on the "Actions" tab at the top, then click on the newest workflow run, then on the "plugin" artifact at the bottom and you'll be able to download a ZIP file with the newest plugin packaged from the latest commit.

**Reply by ElleKayEm** [reply]
You mention in the changelog that you don't have B&N PDFs to test.  If you search for PagePerfect and sort by price low to high, there are a few freebies:  https://www.barnesandnoble.com/s/pageperfect?Ns=P_Sale_Price%7C0

**Reply by noDRM** [reply]
Can't do that as I'm not in the US, and they don't seem to let me "buy" a free PDF without an US credit card and an US billing address.

**Reply by ElleKayEm** [reply]
Ah, I see.  When I get a chance, I will post a log of my attempt to decrypt a B&N PDF.

**Reply by spoolio** [reply]
@charles-wilt, noDRM;
  Seeing same on 5.34/64bit. Nov release works normally. Thanks gents--you guys are fantastic!

calibre Debug log
calibre 5.34 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: Obok DeDRM (10, 0, 0)
calibre 5.34 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: Obok DeDRM (10, 0, 0)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 108.37333333333332 x 107.5764705882353
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[0.45] splash screen shown
[0.45] Initializing db...
[0.45] db initialized
[0.45] Constructing main UI...
DEBUG:    0.0 obok::utilities.py - loading translations
DEBUG:    0.0 obok::dialogs.py - loading translations
DEBUG:    0.0 obok::config.py - loading translations
DEBUG:    0.0 obok::action_err.py - loading translations
[0.99] main UI initialized...
[0.99] Hiding splash screen
Starting QuickView
calibre 5.34 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (10, 0, 2) && Obok DeDRM (10, 0, 0)
Traceback (most recent call last):
  File "calibre\gui2\preferences\plugins.py", line 303, in customize_plugin
  File "calibre\gui2\preferences\plugins.py", line 381, in modify_plugin
  File "calibre\customize\__init__.py", line 166, in do_user_config
  File "calibre_plugins.dedrm.__init__", line 1042, in config_widget
ModuleNotFoundError: No module named 'config'

[8.92] splash screen hidden
[8.92] Started up in 8.92 seconds with 1 books
stdout+stderr from file dialog helper: [b'', b'']

**Reply by noDRM** [reply]
Okay, that probably means I broke something when I began working on the standalone version. I will try to get this fixed tomorrow.

**Reply by noDRM** [reply]
This issue should now hopefully be fixed.

**Reply by agentxan** [reply]
Using the latest I get this error

DeDRM v10.0.2: Trying to decrypt 9780446506342.epub
DeDRM v10.0.2: Verifying zip archive integrity
DeDRM v10.0.2: “9780446506342.epub” is a secure PassHash-protected (B&N) ePub
DeDRM v10.0.2: Trying Encryption key nook_nookStore_X
zk2zxnhh.epub is a secure Adobe PassHash (B&N) ePub.
Could not decrypt zk2zxnhh.epub. Wrong key
DeDRM v10.0.2: Failed to decrypt with key nook_nookStore_X after 0.1 seconds
DeDRM v10.0.2: Looking for new NOOK Keys after 0.1 seconds
No nook Study log files have been found.
DeDRM v10.0.2: Exception when getting PassHashes from ADE after 0.2 seconds
Traceback (most recent call last):
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\adobekey_get_passhash.py", line 75, in passhash_keys
FileNotFoundError: [WinError 2] The system cannot find the file specified

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 414, in ePubDecrypt
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\adobekey_get_passhash.py", line 77, in passhash_keys
adobekey_get_passhash.ADEPTError: Could not locate ADE activation
[2.45] splash screen hidden
[2.45] Started up in 2.45 seconds with 3 books
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 1035, in run
  File "calibre_plugins.dedrm.__init__", line 434, in ePubDecrypt
UnboundLocalError: local variable 'defaultkeys_ade' referenced before assignment

[tmpqmkto8r0.txt](https://github.com/noDRM/DeDRM_tools/files/7798141/tmpqmkto8r0.txt)

If I go back to the official release it completes fine.

dded Across the Sea of Suns to db in: 0.1
Added 1 books in 0.9 seconds
DeDRM v10.0.2: Trying to decrypt 9780446506342.epub
DeDRM v10.0.2: Verifying zip archive integrity
DeDRM v10.0.2: “9780446506342.epub” is a secure Barnes & Noble ePub
DeDRM v10.0.2: Trying Encryption key nook_nookStore_X
Watermark: Successfully stripped 81 ADEPT watermark(s) from ebook.
DeDRM v10.0.2: Post-processing took 0.3 seconds
DeDRM v10.0.2: Finished after 0.5 seconds

**Reply by noDRM** [reply]
@agentxan That error (`FileNotFoundError: [WinError 2] The system cannot find the file specified`) occurs because it fails to decrypt with the key, then tries to find new keys in various places and fails because you don't have ADE installed (which can also contain PassHash keys sometimes). I will get that fixed tomorrow so that on machines without ADE that error no longer occurs.
Though I wonder why the first decryption (before searching for new keys) didn't work ...

Assuming the `nook_nookStore_X` key is the same one in both of your tests, I don't see any reason why the new version would fail to decrypt the book but the old one would work. The only thing that's changed (in theory) is the key retrieval code ...
I'll look over my code again to see if I broken something.

Can you check the dedrm.json file in the plugins folder and see if the "nook_nookStore_X" key is identical with both versions?

@charles-wilt If you're also getting decryption / padding errors with the standalone python script I doubt that that would work with the plugin. The books you tried to remove DRM from are from the same machine where you got the key from? Did you try redownloading the books?

**Reply by agentxan** [reply]
yes the nook_nookStore_X key is the same.  

I get a successful decryption if in verify_book_key(bookkey) I comment out

    #if bookkey[-17] != '\x00' and bookkey[-17] != 0:
    #    # Byte not null, invalid result
    #    return False

    #if ((bookkey[0] != '\x02' and bookkey[0] != 2) and
    #    ((bookkey[0] != '\x00' and bookkey[0] != 0) or 
    #    (bookkey[1] != '\x02' and bookkey[1] != 2))):
    #    # Key not starting with "00 02" or "02" -> error
    #    return False

**Reply by noDRM** [reply]
Oops, looks like I made a mistake when integrating the B&N code into the ineptepub.py script. No idea why I didn't catch that during my testing. 

Can you try the new bugfix (just download again from the latest Actions run) and see if it's now working again?

**Reply by agentxan** [reply]
The latest build also works for me but if I remove the key and try to add a book it finds and adds the key and removes the drm but it gives the following error.

DeDRM v10.0.2: Trying to decrypt 9780446506342.epub
DeDRM v10.0.2: Verifying zip archive integrity
DeDRM v10.0.2: “9780446506342.epub” is a secure PassHash-protected (B&N) ePub
DeDRM v10.0.2: Looking for new NOOK Keys after 0.1 seconds
No nook Study log files have been found.
DeDRM v10.0.2: Exception when getting PassHashes from ADE after 0.2 seconds
Traceback (most recent call last):
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\adobekey_get_passhash.py", line 75, in passhash_keys
FileNotFoundError: [WinError 2] The system cannot find the file specified

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 416, in ePubDecrypt
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\adobekey_get_passhash.py", line 77, in passhash_keys
adobekey_get_passhash.ADEPTError: Could not locate ADE activation
DeDRM v10.0.2: Trying a new default key
q4rmob51.epub is a secure Adobe PassHash (B&N) ePub.
DeDRM v10.0.2: Saving a new default key
DeDRM v10.0.2: Saved a new default key after 0.3 seconds
Watermark: Successfully stripped 81 ADEPT watermark(s) from ebook.
DeDRM v10.0.2: Post-processing took 0.3 seconds
DeDRM v10.0.2: Finished after 0.6 seconds
[2.78] splash screen hidden
[2.78] Started up in 2.78 seconds with 3 books
Added Across the Sea of Suns to db in: 0.1
Added 1Worker Launch took: 0.00 seconds

**Reply by noDRM** [reply]
Thanks, that should now be fixed as well.

**Reply by agentxan** [reply]
yes that fixed it. thank you

**Comment by spoolio** [comment]
Good results with fesiwi's method! Exfiltrated my spouse's--OK, a few were mine ;)--Nook collection so we can read on our local e-ink devices vice crappy phone screens. 

Windows 10 10.0.19041
Calibre 5.34 64bit
Nook app 1.11.0.4


  For those Google-surfers more on my level,  I just coped the terminal output from fesiwi's method above into a notepad terminal and saved with the .b64 extension.  I imported that file using the "import Existing Keyfiles" button on the Barnes and Noble customization control tool under  Prefs>Plugins>(highlight DeDRM)>Customize Plugin>Barnes and Noble ebooks.  I imported all the downloaded Nook files from the app at:   %user%\AppData\Local\Packages\BarnesNoble.Nook_ahnzqzva31enc\LocalState\  via the Calibre Add Books tool, and they all showed up good-to-go on in my Calibre Library.

  Fesiwi, I'm not sure how that code works--but I'm glad you're on our side :)

J

**Comment by PKUD1544** [comment]
noDRM, fesiwi,and spoolio
Thank You to all 3 of you.  With your instructions, I was finally able to liberate 6 years of B&N book purchases (My nook ST finally died a horrible death) and transfer them to my newish PW4

**Comment by nh905** [comment]
I have been using DeDRM 7.21 on Calibre 5.33.2 to remove DRM from Barnes&Noble books so that I can use a decent epub reader.  I have three B&N keys configured in DeDRM, the latest retrieved in 2018.  I downloaded a book and a magazine yesterday - DRM removal was successful on the magazine but not the book (none of the keys worked).  DeDRM was not able to fetch/create new keys ("Could not retrieve key").  ElleKayEm pointed me here.  

I have Nook app version 1.11.0.4 installed on Windows 10 Home and can use it to read the book.  I downloaded and unzipped nook-decoder.py from the top post into the Calibre2 directory.  When I run the latest nook-decoder.py, I get:

> C:\Program Files\Calibre2>calibre-debug -e nook-decoder.py
Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\debug.py", line 308, in main
  File "calibre\debug.py", line 243, in run_script
  File "polyglot\builtins.py", line 110, in exec_path
  File "C:\Program Files\Calibre2\nook-decoder.py", line 57, in <module>
    main()
  File "C:\Program Files\Calibre2\nook-decoder.py", line 53, in main
    cc_hash = unpad(AES.new(hash_key, AES.MODE_CBC, encrypted_cc_hash[:16]).decrypt(encrypted_cc_hash[16:]), 16)
  File "Crypto\Util\Padding.py", line 90, in unpad
ValueError: Padding is incorrect.

Has B&N tightened the screws again or should I have first replaced DeDRM with the latest noDRM?  

Thanks, Norbert

**Reply by ElleKayEm** [reply]
Shouldn't matter whether you have Harper or noDRM's DeDRM as far as running nook-decoder goes.  Just running it on the machine where you have calibre and the Nook app.

**Reply by nh905** [reply]
I dumped the serial_number, hash_key, and the first set of pass_hash/encrypted_cc_hash values before nook-decoder.py fails in line 53 by adding the following statements - I am not sure str() is the correct conversion:

> print("serial_number=" + serial_number)
print("hash_key=" + str(hash_key))
print("pass_hash=" + pass_hash.text)
print("encrypted_cc_hash=" + str(encrypted_cc_hash))

However, I have no clue what I am looking at.  If none of these values contain sensitive data, I can post the results here.

**Reply by nh905** [reply]
I had three keys defined in DeDRM, one from 2018 and two even older ones.  The activation.xml file contained three <adept:passHash> entries.  Since nook-decoder.py failed on the first one, I updated the code to skip the first hash and successfully retrieved the remaining hashes.   The second hash matched my 2018 key but the third hash was new.  I added it to DeDRM and successfully removed DRM from the troublesome book.  My immediate problem is solved but if B&N starts delivering books with the first hash, I will once again be unable to remove the DRM.

I found a NookReaderLog.txt file in the B&N appdata\packages folder.  It contained a number of errors relating to license injection that might be relevant.  So far, no luck with Internet searches on the error.
> Error:  2022-01-11 2:59:32 PM -05:00 ReaderLib c:\build\jenkins\workspace\windows10-x86-release\metro\libs\reader\rmsdk\impls\DocumentImpl.h 272 E_ADEPT_CORE_PASSHASH_NOT_FOUND http://edelivery.barnesandnoble.com/EDS/fulfillment
Error:  2022-01-11 2:59:32 PM -05:00 ReaderLib c:\build\jenkins\workspace\windows10-x86-release\metro\libs\reader\rmsdk\impls\DocumentImpl.h 272 E_ADEPT_CORE_NO_LICENSE
Error:  2022-01-11 2:59:32 PM -05:00 RMSDK ..\..\public\ReaderProvider.cpp 734 Failed to create RMSDK document file:///C:/Users/Norbert/AppData/Local/Packages/BarnesNoble.Nook_ahnzqzva31enc/LocalState/9781541619296 application/epub+zip
Error:  2022-01-11 2:59:32 PM -05:00 RMSDK ..\..\public\ReaderProvider.cpp 748 Error during license injection: now try to delete activation.xml ...
Error:  2022-01-11 2:59:32 PM -05:00 RMSDK ..\..\public\ReaderProvider.cpp 785 Error during license injection E_ADEPT_CORE_NO_LICENSE

**Reply by noDRM** [reply]
Interesting. I doubt that the non-working hash is part of a DRM update by B&N, I think it's more likely that it just got corrupted somehow. I'm going to update the plugin's implementation of the code to just ignore errors when parsing a single entry so that it will read the whole list and add all keys it can decrypt.

**Reply by noDRM** [reply]
Took a while, but that's now implemented in the plugin as well - if it fails to parse one passHash, it will just skip that and still attempt to read all the others.

**Comment by junefish** [comment]
> Experimental Python script to extract the Barnes and Noble Nook decryption key (ccHash) from the Nook Windows store app. If it proves useful perhaps it could be included into the DeDRM plugin. The actual script has been zipped for posting here.
> 
> Instructions:
> 
>     1. Install calibre version 5 or later on your PC. (This Python script relies on calibre version 5 or later being installed to supply
>        the correct libraries for its execution.)
> 
>     2. Install the Nook app from: https://www.microsoft.com/en-us/p/nook-books-magazines-newspapers-comics/9wzdncrfj33h
>        (Requires a recent Windows version.)
> 
>     3. Sign in to the Nook app with your email and password and then close the app.
> 
>     4. Run this script by entering at a Windows command prompt:
>        calibre-debug -e nook-decoder.py
> 
> 
> This procedure has been tested with Nook app version 1.11.0.4 under Windows 11.
> 
> [nook-decoder.zip](https://github.com/noDRM/DeDRM_tools/files/7740134/nook-decoder.zip)

This worked perfectly for me!

**Reply by dslaby** [reply]
Running the script fails with the following results:
![nook-decoder output](https://user-images.githubusercontent.com/3310619/209453334-e70410b8-2fd6-41ed-b775-e3dc6872c70c.JPG)
What can I do to correct it?

**Reply by ElleKayEm** [reply]
This is built into the plugin now.

**Comment by DARK-MASKS** [comment]
what did I do wrong?

C:\Users\Natha>calibre-debug -e nook-decoder.py
Integration status: True
Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\debug.py", line 308, in main
  File "calibre\debug.py", line 243, in run_script
  File "polyglot\builtins.py", line 107, in exec_path
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Natha\\nook-decoder.py'

**Reply by noDRM** [reply]
The error message tells you exactly what you did wrong - you are executing this command in `C:\Users\Natha`, and that folder doesn't contain the nook-decoder.py script.

**Comment by StarkindlerBE300** [comment]
Hello! Github noob here! 
I feel ignorant asking but after following (I think) the instructions to liberate my library from B&N there is no joy. DRM is still in place. I feel like I missed something simple but with @250 books to free from captivity it seems daunting. A quick check after importing shows about 62%+ of the books are still under DRM. I have the latest Calibre (6.4) with nothing imported at the moment. I also have the M$-Store Nook app installed (1.11.0.4) and all but one book downloaded.
Please help me escape the captivity!

**Reply by ElleKayEm** [reply]
Don't try to import them all at once or DeDRM will get overwelmed and some will slip past without decryption.  Try 20 or so and see if all of those worked.  Then you can try more or less with the next batch depending on how it goes.

**Comment by StarkindlerBE300** [comment]
Thanks! I will give smaller batches a try.

**Reply by StarkindlerBE300** [reply]
Thanks a ton!! Using groups of 5 at a time I was able to get 128 out of the first 141 books to import/convert properly (i checked each one to be sure it opened).
 Then they all started to fail. I shut down for a bit and started again, but they still all fail now. (??)

**Reply by ElleKayEm** [reply]
Were some of them downloaded a while ago?  Maybe try a fresh download.  It's usually all or nothing.  But perhaps your key changed since some of them were downloaded.

**Comment by StarkindlerBE300** [comment]
Thank you again.
I had thought of that and erased all, uninstalled the Nook reader app, then
reinstalled and re-downloaded all the books so they were “fresh”.
It’s an odd issue because the books that are staying “locked” seem spread
out over my purchase history but do lean toward older purchases.

There may be an issue with an older key, but the account has been linked to
the same email since I got the first nook on pre-order ages ago.

**Reply by ElleKayEm** [reply]
I know my key hasn't changed since they switched away from credit card based keys back in 2014.  But others have had them change occasionally.  Let's take a look at an import log of just one problem book: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md#i-cannot-solve-my-problem-with-the-dedrm-plugin-and-now-i-need-to-post-a-log-how-do-i-do-that

**Reply by StarkindlerBE300** [reply]
Log pasted below:
=============================================================================

ApplicationPaletteChange event ignored
Using calibre Qt style: True
calibre Debug log
calibre 6.4  embedded-python: True
Windows-10-10.0.19043-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19043')
Python 3.10.1
Windows: ('10', '10.0.19043', 'SP0', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (10, 0, 2)
calibre 6.4  embedded-python: True
Windows-10-10.0.19043-SP0 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19043')
Python 3.10.1
Windows: ('10', '10.0.19043', 'SP0', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (10, 0, 2)
QPA platform: windows
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 92.53889943074003 x 92.67567567567568
[0.00] Starting up...
[0.00] Showing splash screen...
[0.19] splash screen shown
[0.19] Initializing db...
[0.19] db initialized
[0.19] Constructing main UI...
[0.77] main UI initialized...
[0.77] Hiding splash screen
Starting QuickView
DeDRM v10.0.2: Trying to decrypt 9780062101334.epub
DeDRM v10.0.2: Verifying zip archive integrity
DeDRM v10.0.2: “9780062101334.epub” is a secure Barnes & Noble ePub
DeDRM v10.0.2: Looking for new NOOK Study Keys after 0.1 seconds
No nook Study log files have been found.
DeDRM v10.0.2: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 178, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 899, in run
  File "calibre_plugins.dedrm.__init__", line 397, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v10.0.2: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
[15.84] splash screen hidden
[15.84] Started up in 15.84 seconds with 128 books
stdout+stderr from file dialog helper: [b'', b'']
piped data from file dialog helper: [b'\xff\x83o_\x1d\xbf\xb4\xee\xaeE\x17*\xc9\x05nN\xcd\x02 \xf1\x85\xc9\xa2#\x83}\xf9\xebJ$<E', b'C:\\Users\\david\\AppData\\Local\\Packages\\BarnesNoble.Nook_ahnzqzva31enc\\LocalState\\not good\\9780062101334.epub']
Added The Other Side of Normal to db in: 4.1

================================================================================
I can usually tell the import is not goo because there is no cover displayed. not always, but 90%+ of the time. I have the growing list of "bad" books in a folder named "not good" so I can keep track of them. the import fails no matter what folder they are in.

**Reply by StarkindlerBE300** [reply]
Question? would the Nook reader app need to be open? I have been importing with it closed down after getting the key.

**Reply by ElleKayEm** [reply]
With calibre 6, use DeDRM 10.0.3.

**Reply by ElleKayEm** [reply]
No, the Nook app doesn't need to be open.

**Reply by StarkindlerBE300** [reply]
Well Thanks again! didn't know I had to find 10.0.3...thought the latest was in the download. I'll search for it later and give it a try.

**Reply by ElleKayEm** [reply]
It's the latest release here.  See this page: https://github.com/noDRM/DeDRM_tools/releases/tag/v10.0.3

**Reply by StarkindlerBE300** [reply]
Thank You !! got it installed and it seems to be working. I will hopefully finish importing them tomorrow. Many Thanks!!

**Reply by StarkindlerBE300** [reply]
...and there was JOY in the midst of  Nook Captivity as the chains fell off...!!
Many thanks !!

**Comment by TheDude005** [comment]
Just curious if this would work.  If install Calibre on a windows machine and use this script to obtain the hash can I then take that hash over to my Ubuntu server and enter it in to the DeDRM plugin there and have it still work?

**Reply by ElleKayEm** [reply]
This script is now built into DeDRM.  But, yes, if you install Windows calibre with DeDRM where you have the Windows Nook app, you should be able to save the key and then import it into your Linux setup.

**Comment by twilight-flower** [comment]
For whatever reason, it seems that this is only working on some books for me, not on all. I've currently got two books downloaded in my Nook app, one 800kB Jane Austen novel and one 80MB manga volume. I've confirmed that both are DRM-laden, within the Nook download directory. And importing the Jane Austen novel into Calibre strips the DRM from it without any fuss, but importing the manga volume into Calibre doesn't; it remains DRM-laden and unusable even post-import. (I'm *guessing* this is a small-file-versus-large-file issue, but I don't own enough Nook books of different file sizes to be able to do any serious experimentation to confirm it; perhaps it's some other thing entirely, and the file-size difference is just a red herring.)

I'm not sure what to try here. Based on the discussions upthread of issues when importing too many books at a time, I wondered if it was a "my CPU is busy being hogged by some other program and thus Calibre is being throttled as though in a multi-book export"-type issue, but I tried again once the other program was done and my CPU was freed up, and the issue persisted. I've failed to generate additional ideas since then. Do people here have any suggestions?

**Reply by ElleKayEm** [reply]
I found with comics downloaded with the Windows Nook app that part of the DRM is stored in a separate database.  I commented on it here: https://github.com/apprenticeharper/DeDRM_tools/issues/693#issuecomment-442184522

I believe I have heard that comics from the Android app or a Nook device have the usual DRM setup, but you need root access to get to the files.

**Reply by twilight-flower** [reply]
That method solved it. Thanks!

**Reply by ElleKayEm** [reply]
Glad to hear it!

**Comment by rcentros** [comment]
Worked great first time out. Thank you. I use Linux but this computer has a Windows 10 install that I shrunk and left on its partition before installing Linux. So I finally found something to use it for. I had tried the Android Studio method but kept coming up with a NULL value when I ran SQLite. I guess the NoDRM method would only work in Windows (which I'm going to test). 

I had "liberated" most of my B&N books when the encryption was based on my CC number. But I've bought a few since. Mostly though, I wanted to get this working because I would rather buy from B&N than Amazon (they often run the same sales). Now I can do that without feeling "trapped."

**Reply by ElleKayEm** [reply]
The [help file](https://github.com/noDRM/DeDRM_tools/blob/master/DeDRM_plugin/DeDRM_ADE%20PassHash%20Key_Help.htm) in DeDRM for ADE PassHash keys also has some info about extracting the key from the Nook Android app.  But for me that key is the same as the one from the Windows app.  I've actually had the same key ever since they switched away from cc based keys.
